src/utils/Block.ts

```diff
@@ -0,0 +1,238 @@
+    this._registerEvents(eventBus);
+    eventBus.emit(Block.EVENTS.INIT)
+  }
+
+  getChildren(propsAndChildren: any) {
+    const children: any = {}
+    const props: any = {}
+
+    Object.entries(propsAndChildren).map(([key, value]) => {
+      if (value instanceof Block) {
+        children[key] = value;
+      } else if (Array.isArray(value) && value.every(v => (v instanceof Block))) {
+        children[key] = value;
+      } else {
+        props[key] = value
+      }
+    });
+    return { props, children }
+  }
+
+
+  _registerEvents(eventBus: EventBus) {
+    eventBus.on(Block.EVENTS.INIT, this.init.bind(this));
+
+    eventBus.on(Block.EVENTS.FLOW_CDM, this._componentDidMount.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDU, this._componentDidUpdate.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_RENDER, this._render.bind(this));
+  }
+
+  //   _createResources() {
+  //     this._element = this._createDocumentElement('div');
+  //   }
+
+  //   protected getStateFromProps(props: any): void {
+  //     this.state = {};
+  //   }
+
+  init() {
+    // this._createResources();
+    this.eventBus().emit(Block.EVENTS.FLOW_RENDER, this.props);
+  }
+
+  _componentDidMount() {
+    this.componentDidMount();
+  }
+
+  componentDidMount() {
+  }
+
+  dispatchComponentDidMount() {
+    this.eventBus().emit(Block.EVENTS.FLOW_CDM);
+  }
+
+  _componentDidUpdate(oldProps: any, newProps: any) {
+    if (this.componentDidUpdate(oldProps, newProps)) {
+      this.eventBus().emit(Block.EVENTS.FLOW_RENDER);
+    }
+  }
+
+  componentDidUpdate(oldProps: any, newProps: any) {
+    return true;
+  }
+
+  setProps = (nextProps: any) => {
+    if (!nextProps) {
+      return;
+    }
+
+    Object.assign(this.props, nextProps);
+  };
+
+
+  get element(): HTMLElement | null {
+    return this._element;
+  }
+
+  _render() {
+    const fragment = this.render();
+    const newElement = fragment.firstElementChild as HTMLElement;
+    if (this._element) {
+      this._removeEvents();
+      this._element.replaceWith(newElement);
+    }
+
+    this._element = newElement;
+
+    this._addEvents();
+
+
+  }
+
+  protected render(): DocumentFragment {
+    return new DocumentFragment();
+  }
+
+  getContent(): HTMLElement | null {
+    return this.element;
+  }
+
+  _makePropsProxy(props: any): any {
+    // Можно и так передать this
+    // Такой способ больше не применяется с приходом ES6+
+    const self = this;
+
+    return new Proxy(props as unknown as object, {
+      get(target: Record<string, unknown>, prop: string) {
+        const value = target[prop];
+        return typeof value === 'function' ? value.bind(target) : value;
+      },
+      set(target: Record<string, unknown>, prop: string, value: unknown) {
+        target[prop] = value;
+
+        // Запускаем обновление компоненты
+        // Плохой cloneDeep, в след итерации нужно заставлять добавлять cloneDeep им самим
+        self.eventBus().emit(Block.EVENTS.FLOW_CDU, { ...target }, target);
```

Comment:
Можно использовать стрелочные функции, тогда не придется прибегать к замыканию
```suggestion
    // Можно и так передать this
    // Такой способ больше не применяется с приходом ES6+
    const self = this;

    return new Proxy(props as unknown as object, {
      get(target: Record<string, unknown>, prop: string) {
        const value = target[prop];
        return typeof value === 'function' ? value.bind(target) : value;
      },
      set: (target: Record<string, unknown>, prop: string, value: unknown) => {
        target[prop] = value;

        // Запускаем обновление компоненты
        // Плохой cloneDeep, в след итерации нужно заставлять добавлять cloneDeep им самим
        this.eventBus().emit(Block.EVENTS.FLOW_CDU, { ...target }, target);
```